### PR TITLE
[16.0] [ADD] shopinvader_product_sale_packaging

### PR DESCRIPTION
**Pull request by marielejeune**
Add packaging info on products exports. Extracted from `shopinvader_sale_packaging`

NB: When migrating from v14 to v16 model `product.packaging.type` was renamed `product.packaging.level`

### CODE REVIEW

File: shopinvader_product_sale_packaging/models/product_product.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+        }
+
+    def _prepare_qty_by_packaging_values(self, packaging, qty_per_pkg):
+        return {
+            "id": packaging.id,
+            "qty": qty_per_pkg,
+            "name": packaging.name,
+            "is_unit": packaging.is_unit,
+            "barcode": packaging.barcode,
+        }
```
Comment (marielejeune) [inline shopinvader_product_sale_packaging/models/product_product.py]: Not sure to understand why this method is redefined here as it is a duplicate of stock_packaging_calculator/models/product.py (L147), already in v14.

File: shopinvader_product_sale_packaging/models/product_product.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+                pkg_info["can_be_sold"] = not self.sell_only_by_packaging
+            res.append(pkg_info)
+        return res
+
+    def _get_variant_packaging_ctx(self):
+        return {
+            # consider only packaging that can be displayed
+            "_packaging_filter": lambda x: x.shopinvader_display,
+            "_packaging_values_handler": self._prepare_qty_by_packaging_values,
+        }
```
Comment (marielejeune) [inline shopinvader_product_sale_packaging/models/product_product.py]: In v14 `lang` was given in context, from the backend. How to manage it now?

File: shopinvader_product_sale_packaging/views/product_packaging.xml [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    <record id="product_packaging_tree_view_inherit" model="ir.ui.view">
+        <field name="model">product.packaging</field>
+        <field
+            name="inherit_id"
+            ref="sell_only_by_packaging.product_packaging_tree_view_inherit"
+        />
+        <field name="arch" type="xml">
+            <field name="force_sale_qty" position="after">
+                <!-- TODO: group on this field was shopinvader.group_shopinvader_manager-->
+                <field name="shopinvader_display" />
```
Comment (marielejeune) [inline shopinvader_product_sale_packaging/views/product_packaging.xml]: How to manage security now that this module shouldn't depend on `shopinvader_restapi` 
 anymore?

Comment (marielejeune) [reply shopinvader_product_sale_packaging/views/product_packaging.xml]: Adding "Product / manager" group (base.group_system) for now.

File: shopinvader_product_sale_packaging/models/product_product.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+            x["id"]: x["can_be_sold"] for x in self.packaging_ids.read(["can_be_sold"])
+        }
+        for pkg in packaging:
+            pkg_info = self._prepare_qty_by_packaging_values(pkg, pkg.qty)
+            pkg_info["contained"] = contained_mapping.get(str(pkg.id))
+            pkg_info["can_be_sold"] = can_be_sold_info.get(pkg.id, False)
+            if pkg.is_unit:
+                pkg_info["can_be_sold"] = not self.sell_only_by_packaging
+            res.append(pkg_info)
+        return json.dumps(res)
```
Comment (marielejeune) [inline shopinvader_product_sale_packaging/models/product_product.py:51]: In v14 we returned `res`, but this is a list, which is not supported by `base_sparse_field` anymore.
This solution is not working either. Idk what to do ...

File: shopinvader_product_sale_packaging/models/product_product.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+        }
+        for pkg in packaging:
+            pkg_info = self._prepare_qty_by_packaging_values(pkg, pkg.qty)
+            pkg_info["contained"] = contained_mapping.get(str(pkg.id))
+            pkg_info["can_be_sold"] = can_be_sold_info.get(pkg.id, False)
+            if pkg.is_unit:
+                pkg_info["can_be_sold"] = not self.sell_only_by_packaging
+            pkg_id = pkg_info.pop("id")
+            res[pkg_id] = pkg_info
+        return res
```
Comment (marielejeune) [inline shopinvader_product_sale_packaging/models/product_product.py]: This `res` is stored in a Serialized field. In v16 the previous code (returning a list of dicts) is not working anymore. (`TypeError: the JSON object must be str, bytes or bytearray, not list`)
The only solution I see is to update a bit the structure, returning a dict of dicts.
But it changes the structure ... What do you think? Shouldn't we trigger a recompute of the field then?
@lmignon

**Comment by marielejeune** [comment]
@sbidoul @lmignon tests of `shopinvader_anonymous_partner` are failing, do you have an idea why? 
This addon shouldn't be concerned by this PR...

File: shopinvader_product_sale_packaging/models/product_product.py (hunk 5 above)
Comment (marielejeune) [reply shopinvader_product_sale_packaging/models/product_product.py]: Removing this field and exposing packaging as a list of ProductPackaging Pydantic schemas.

**Comment by marielejeune** [comment]
@sbidoul `test_create_and_get` in `shopinvader_anonymous_partner` is failing because of this Odoo bug: https://github.com/odoo/odoo/pull/137997

Do you see a way to make this pipeline succeed? The company is lost from the context when calling `url_open`.

**Comment by sbidoul** [comment]
:green_apple:

**Comment by simahawk** [comment]
@marielejeune all deps got merged ;)

File: shopinvader_product_sale_packaging/schemas/product_packaging.py [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+            contained_packaging = packaging_contained_mapping.get(str(obj.id), [])
+            if contained_packaging:
+                obj.contained = [
+                    SimpleProductPackaging.from_packaging(
+                        odoo_product,
+                        Packaging(
+                            pkg["id"],
+                            pkg["name"],
+                            pkg["qty"],
+                            pkg["barcode"],
```
Comment (lmignon) [inline shopinvader_product_sale_packaging/schemas/product_packaging.py]:
```suggestion
                            pkg["barcode"] or None,
```

**Comment by simahawk** [comment]
can we merge or extrapolate the fix by @sbidoul ?

**Pull request merged**